### PR TITLE
Serve client files from additional routes

### PR DESCRIPTION
**Pull request by nwalters512**
This work is a prerequisite for moving course content onto subdomains.

The main change here: when client files are served for assessments or questions, they're served from a path that includes the question / assessment ID (or, for student pages, the question instance or assessment instance ID). This ensures that we can reliably serve those routes from the same subdomain as we serve the question itself, which is important to avoid CORS issues.

Note that the "global" routes for course and course instance client files remain, as instructors are in some cases hardcoding paths to those routes. So, we have to keep them around for backwards compatibility.

### CODE REVIEW

File: apps/prairielearn/src/lib/question.js [hunk 1]
```diff
@@ -1399,9 +1399,10 @@ module.exports = {
       urls.calculationQuestionFileUrl = questionUrl + 'file';
 
       // FIXME: broken?
-      urls.calculationQuestionGeneratedFileUrl = questionUrl + 'generatedFilesQuestion';
+      urls.calculationQuestionGeneratedFileUrl =
+        questionUrl + 'generatedFilesQuestion/variant/' + variant.id;
 
-      urls.clientFilesCourseUrl = urlPrefix + '/clientFilesCourse';
+      urls.clientFilesCourseUrl = questionUrl + 'clientFilesCourse';
```
Comment (mwest1066) [inline apps/prairielearn/src/lib/question.js:1405]: Feels like we are missing a `/` here? Also in the line below in this file.

File: apps/prairielearn/src/server.js [hunk 2]
```diff
@@ -1430,15 +1450,32 @@ module.exports.initExpress = function () {
     );
   }
 
-  // clientFiles
+  // Global client files
   app.use('/pl/course_instance/:course_instance_id/clientFilesCourse', [
+    // TODO: this is no-op? remove here and elsewhere?
```
[the remark this answers is not in the document]
Comment (mwest1066) [reply apps/prairielearn/src/server.js:1455]: I agree that `studentAssessmentAccess` can't be doing anything useful for routes that don't have an assessment so it should be safe to remove here.

**Review by mwest1066 — APPROVED** [review]

**Comment by mwest1066** [comment]
Do we have any tests for these kind of routes? I'm slightly nervous that our tests might never be loading client files.

File: apps/prairielearn/src/lib/question.js (hunk 1 above)
Comment (nwalters512) [reply apps/prairielearn/src/lib/question.js:1405]: Look a few lines up - `questionUrl` is defined with a trailing slash. Ditto for below.

File: apps/prairielearn/src/server.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (nwalters512) [reply apps/prairielearn/src/server.js:1455]: I opened https://github.com/PrairieLearn/PrairieLearn/pull/8399 to clean this up.

**Comment by nwalters512** [comment]
> Do we have any tests for these kind of routes?

Good call: as far as I can tell, we don't! I can add some and hopefully hit all these routes.

**Comment by nwalters512** [comment]
I added tests for course and course instance client files in b494ee3. I skipped assessment and question files for now, as we don't have good primitives for creating arbitrary assessment instances and instance questions (or our existing primitives aren't flexible enough to support adding more questions.

The good news: there were in fact some existing tests for question files 🎉

File: apps/prairielearn/src/lib/question.js (hunk 1 above)
Comment (mwest1066) [reply apps/prairielearn/src/lib/question.js:1404]: Ah! Apologies, I had misread a wrapped line.

**Comment by mwest1066** [comment]
> I added tests for course and course instance client files in [b494ee3](https://github.com/PrairieLearn/PrairieLearn/commit/b494ee3e7d65500e2d2edc71d477bf17335d27e5). I skipped assessment and question files for now, as we don't have good primitives for creating arbitrary assessment instances and instance questions (or our existing primitives aren't flexible enough to support adding more questions.
> 
> The good news: there were in fact some existing tests for question files 🎉

Awesome! Many thanks for the new tests 🙏

**Review by mwest1066 — APPROVED** [review]

**Comment by nwalters512** [comment]
I separated the usage of these routes into #8470 to make it slightly safer to deploy this change.

**Pull request merged**